From 57932327babe39308e158739e679ba3e92bb15e2 Mon Sep 17 00:00:00 2001
From: Sankalp Negi <snegi@qti.qualcomm.com>
Date: Thu, 20 Feb 2025 04:53:53 -0800
Subject: [PATCH] PENDING: pinctrl: qcom: Enable read-after-write feature

Add Functional Safety (FuSa) framework support for
read-after-write scenario.

Change-Id: Icb1b7247db17c2c3111f12b1833de3398d857ff0
Signed-off-by: Sankalp Negi <snegi@qti.qualcomm.com>
---
 drivers/pinctrl/qcom/pinctrl-msm.c | 106 ++++++++++++++++++++++++++++-
 1 file changed, 105 insertions(+), 1 deletion(-)

diff --git a/drivers/pinctrl/qcom/pinctrl-msm.c b/drivers/pinctrl/qcom/pinctrl-msm.c
index 6df46f4c968c..4aa809b44c89 100644
--- a/drivers/pinctrl/qcom/pinctrl-msm.c
+++ b/drivers/pinctrl/qcom/pinctrl-msm.c
@@ -21,6 +21,9 @@
 #include <linux/slab.h>
 #include <linux/spinlock.h>
 #include <linux/suspend.h>
+#include <linux/debugfs.h>
+#include <linux/iopoll.h>
+#include <linux/poll.h>
 
 #include <linux/pinctrl/machine.h>
 #include <linux/pinctrl/pinconf-generic.h>
@@ -81,8 +84,13 @@ struct msm_pinctrl {
 	void __iomem *regs[MAX_NR_TILES];
 	u32 phys_base[MAX_NR_TILES];
 	struct msm_gpio_regs *gpio_regs;
+	u32 fusa_read_after_write_errs;
+	bool fusa_read_after_write_triggered;
+	raw_spinlock_t fusa_lock;
+	wait_queue_head_t fusa_wq;
 };
 
+#define IO_REG_BITMASK	~BIT(0)
 #define MSM_ACCESSOR(name) \
 static u32 msm_readl_##name(struct msm_pinctrl *pctrl, \
 			    const struct msm_pingroup *g) \
@@ -92,7 +100,28 @@ static u32 msm_readl_##name(struct msm_pinctrl *pctrl, \
 static void msm_writel_##name(u32 val, struct msm_pinctrl *pctrl, \
 			      const struct msm_pingroup *g) \
 { \
+	int ret; \
+	u32 new_val; \
+	unsigned long flags; \
 	writel(val, pctrl->regs[g->tile] + g->name##_reg); \
+	if (in_atomic() || irqs_disabled() || !preemptible()) { \
+		ret = readl_relaxed_poll_timeout_atomic(pctrl->regs[g->tile] + \
+						g->name##_reg, new_val, \
+						val == new_val, 1, 1000); \
+	} else { \
+		ret = readl_relaxed_poll_timeout(pctrl->regs[g->tile] + \
+						g->name##_reg, new_val, \
+						val == new_val, 1, 1000); \
+	} \
+	if (ret) { \
+		if (!strcmp(#name, "io") && ((val & IO_REG_BITMASK) == (new_val & IO_REG_BITMASK))) \
+			return; \
+		raw_spin_lock_irqsave(&pctrl->fusa_lock, flags); \
+		pctrl->fusa_read_after_write_errs++; \
+		pctrl->fusa_read_after_write_triggered = true; \
+		wake_up_interruptible_poll(&pctrl->fusa_wq, EPOLLIN | EPOLLRDNORM); \
+		raw_spin_unlock_irqrestore(&pctrl->fusa_lock, flags); \
+	} \
 }
 
 MSM_ACCESSOR(ctl)
@@ -1640,6 +1669,75 @@ SIMPLE_DEV_PM_OPS(msm_pinctrl_dev_pm_ops, msm_pinctrl_suspend,
 
 EXPORT_SYMBOL(msm_pinctrl_dev_pm_ops);
 
+
+#ifdef CONFIG_DEBUG_FS
+static ssize_t fusa_read_after_write_errs_read(struct file *file,
+						char __user *user_buf,
+						size_t count, loff_t *ppos)
+{
+	struct msm_pinctrl *pctrl = file->private_data;
+	char buf[32];
+	int len;
+
+	len = snprintf(buf, sizeof(buf), "%u\n",
+			pctrl->fusa_read_after_write_errs);
+	return simple_read_from_buffer(user_buf, count, ppos, buf, len);
+}
+
+static __poll_t fusa_read_after_write_errs_notify(struct file *file,
+						   poll_table *wait)
+{
+	struct msm_pinctrl *pctrl = file->private_data;
+	unsigned long flags;
+	__poll_t mask = 0;
+
+	poll_wait(file, &pctrl->fusa_wq, wait);
+	raw_spin_lock_irqsave(&pctrl->fusa_lock, flags);
+	if (pctrl->fusa_read_after_write_triggered) {
+		pctrl->fusa_read_after_write_triggered = false;
+		mask = EPOLLIN | EPOLLRDNORM;
+	}
+	raw_spin_unlock_irqrestore(&pctrl->fusa_lock, flags);
+	return mask;
+}
+
+static const struct file_operations fusa_read_after_write_errs_fops = {
+	.owner = THIS_MODULE,
+	.read = fusa_read_after_write_errs_read,
+	.open = simple_open,
+	.poll = fusa_read_after_write_errs_notify,
+};
+
+static int qcom_tlmm_fusa_create_debugfs(struct msm_pinctrl *pctrl)
+{
+	struct dentry *debugfs_dir;
+	struct dentry *file;
+	bool dir_created = false;
+
+	debugfs_dir = debugfs_lookup("tlmm_hw_safety", NULL);
+	if (!debugfs_dir) {
+		debugfs_dir = debugfs_create_dir("tlmm_hw_safety", NULL);
+		if (IS_ERR_OR_NULL(debugfs_dir)) {
+			dev_err(pctrl->dev, "Failed to create debugfs directory\n");
+			return debugfs_dir ? PTR_ERR(debugfs_dir) : -ENOMEM;
+		}
+		dir_created = true;
+	}
+
+	file = debugfs_create_file("read_after_write_errors", 0444, debugfs_dir,
+					pctrl, &fusa_read_after_write_errs_fops);
+	if (!dir_created)
+		dput(debugfs_dir);
+
+	if (IS_ERR_OR_NULL(file)) {
+		dev_warn(pctrl->dev, "Failed to create debugfs file\n");
+		return file ? PTR_ERR(file) : -ENOMEM;
+	}
+
+	return 0;
+}
+#endif
+
 int msm_pinctrl_probe(struct platform_device *pdev,
 		      const struct msm_pinctrl_soc_data *soc_data)
 {
@@ -1660,6 +1758,8 @@ int msm_pinctrl_probe(struct platform_device *pdev,
 					"qcom,ipq8064-pinctrl");
 
 	raw_spin_lock_init(&pctrl->lock);
+	raw_spin_lock_init(&pctrl->fusa_lock);
+	init_waitqueue_head(&pctrl->fusa_wq);
 
 	if (soc_data->tiles) {
 		for (i = 0; i < soc_data->ntiles; i++) {
@@ -1708,9 +1808,13 @@ int msm_pinctrl_probe(struct platform_device *pdev,
 
 	platform_set_drvdata(pdev, pctrl);
 
+#ifdef CONFIG_DEBUG_FS
+	ret = qcom_tlmm_fusa_create_debugfs(pctrl);
+#endif
+
 	dev_dbg(&pdev->dev, "Probed Qualcomm pinctrl driver\n");
 
-	return 0;
+	return ret;
 }
 EXPORT_SYMBOL(msm_pinctrl_probe);
 
-- 
2.34.1

